File: lib/pieces/elephant.rb
require_relative "./piece"

class Elephant < Piece

  def get_moves
    [:up, :down].each do |vertical|
      next unless starting_square.send(vertical) &&
        starting_square.send(vertical).send(vertical) &&
        behind_river?(starting_square.send(vertical).send(vertical))
      [:left, :right].each do |horizontal|
        next unless starting_square.send(horizontal) &&
          starting_square.send(horizontal).send(horizontal)
        blocking_square = starting_square.send(vertical).send(horizontal)
        next if @position.find blocking_square.coordinates
        target_square = blocking_square.send(vertical).send(horizontal)
        target_piece = @position.find target_square.coordinates
        next if target_piece && target_piece.color == @color
        add_move target_square.coordinates
      end
    end
  end
end

File: src/elephant.js
var Piece = require('./piece.js');

function Elephant(color) {
  var self = new Piece(color);

  self.getMoves = function(position) {
    var moves = [];
    var continent = position.BOARD.continents[this.color];
    var directions = [
      ['left', 'up'],
      ['up', 'right'],
      ['right', 'down'],
      ['down', 'left']
    ];
    for (var dir_index = 0; dir_index < 4; dir_index += 1) {
      var direction = directions[dir_index];
      if (!this.square[direction[0]] || !this.square[direction[1]]) {
        continue;
      }
      var blocking_square = this.square[direction[0]][direction[1]];
      if (position[blocking_square.coordinates]) {
        continue;
      }
      if (!blocking_square[direction[0]] || !blocking_square[direction[1]]) {
        continue;
      }
      var target_square = blocking_square[direction[0]][direction[1]];
      if (!continent[target_square.coordinates]) {
        continue;
      }
      if (!position[target_square.coordinates] ||
        position[target_square.coordinates].color !== this.color) {
        moves.push(target_square.coordinates);
      }
    }
    return moves;
  };

  return self;
}

Elephant.prototype.constructor = Elephant;

module.exports = Elephant;
